File: BiosensorsPackage.java
// Here is an example of a concrete package (i.e. that implements the Package
// abstract class)
public class BioSensorsPackage extends Package {
  // Define the constructor
  public BioSensorsPackage(movieData owner) {
    super(owner, 'BioSensorsPackage')
      }
	
  // Following are the 3 abstract methods every package needs to implement.
	
  // Sanity check (See Package.m for more details)
  public void sanityCheck(boolean full) throws Exception {
		
    // If anything relative to the package itself needs to be checked before
    // we check the processes, do it here (throw an exception if needed)
    // ...
		
    // check the status of each process, the possible changed of their
    // parameters, and their dependencies with each other.
    checkProcesses_(full);
		
    // If anything relative to the package itself needs to be checked after
    // we've checked the processes, do it here (throw an exception if needed)
    // ...		
  }
	
  // Get the list of process class names required by the package.
  public String[] getListProcessClassName() {
		
    String[] list = new String[3];
		
    list[0] = new String("SegmentationProcess");
    list[1] = new String("Other1Process");
    list[2] = new String("Other2Process");
		
    return list;
  }
	
  // This method return the dependency matrix of processes. It is a square
  // matrix whose size equals the number of required processes. (See Package.m
  // for more details)
  public Matrix getProcessDependenyMatrix() {
    return zeros(3);
  }
}

File: BioSensorsPackage.m
classdef BioSensorsPackage < Package
% A concrete process for BioSensor Package
    
    methods (Access = public)
        function obj = BioSensorsPackage (owner)
           % Construntor of class MaskProcess
           if nargin == 0
              super_args = {};
           else
               % Owner: MovieData object
               super_args{1} = owner;
               super_args{2} = 'BioSensors'; 
               % Dependency Matrix (same length as process class name string)
               %                1 2 3 4 5 6 7 8 9 10
               super_args{3} = [0 0 0 0 0 0 0 0 0 0 ; %1 MasksProcess
                                1 0 0 0 0 0 0 0 0 0 ; %2 BackgroundMasksProcess
                                1 0 0 0 0 0 0 0 0 0 ; %3 MaskRefinementProcess
                                0 0 0 0 0 0 0 0 0 0 ; %4 DarkCurrentCorrectionProcess
                                0 0 0 1 0 0 0 0 0 0 ; %5 ShadeCorrectionProcess
                                0 1 0 0 1 0 0 0 0 0 ; %6 BackgroundSubtractionProcess
                                0 0 0 0 0 1 0 0 0 0 ; %7 BleedthroughCorrectionProcess
                                0 0 0 0 0 1 0 0 0 0 ; %8 TransformationProcess
                                0 0 0 0 0 1 0 0 0 0 ; %9 RatioingProcess
                                0 0 0 0 0 0 0 0 1 0 ];%10PhotobleachCorrectionProcess
                                
               % Process CLASS NAME string (same length as dependency matrix)
               % Must be accurate process class name
               super_args{4} = {'MasksProcess',...              
                                'BackgroundMasksProcess',... 
                                'MaskRefinementProcess',... 
                                'DarkCurrentCorrectionProcess',...
                                'ShadeCorrectionProcess',...
                                'BackgroundSubtractionProcess',...
                                'BleedthroughCorrectionProcess',...
                                'TransformationProcess',...
                                'RatioingProcess',...    
                                'PhotobleachCorrectionProcess',...
                                };
           end
           % Call the supercalss constructor with empty cell array (no
           % argument) if nargin == 0
           obj = obj@Package(super_args{:});
        end
        function procEx = sanityCheck(obj,full,procID) % throws Exception Cell Array
            % Sanity Check
            % full: true or false
            if nargin < 2
                full = true;
            end
            procEx = obj.checkProcesses(full,procID);  % throws Exception Cell Array
        end
    end
    methods (Static)
        function text = getHelp(obj)
            % Function return a string of help text
            text = 'This is specific help text of biosensors package.';
        end
    end
end

